metaflow/datatools/s3.py

```diff
@@ -609,35 +894,43 @@ def _one_boto_op(self, op, url):
     # NOTE: re: _read_many_files and _put_many_files
     # All file IO is through binary files - we write bytes, we read
     # bytes. All inputs and outputs from these functions are Unicode.
-    # Conversion between bytes and unicode is done through url_quote
+    # Conversion between bytes and unicode is done through 
     # and url_unquote.
-    def _read_many_files(self, op, prefixes, **options):
-        prefixes = list(prefixes)
+    def _read_many_files(self, op, prefixes_and_ranges, **options):
+        prefixes_and_ranges = list(prefixes_and_ranges)
         with NamedTemporaryFile(dir=self._tmpdir,
                                 mode='wb',
                                 delete=not debug.s3client,
                                 prefix='metaflow.s3.inputs.') as inputfile:
-            inputfile.write(b'\n'.join(map(url_quote, prefixes)))
+            inputfile.write(b'\n'.join(
+                [b' '.join([url_quote(prefix)] + ([url_quote(r)] if r else []))
+                    for prefix, r in prefixes_and_ranges]))
             inputfile.flush()
             stdout, stderr = self._s3op_with_retries(op,
                                                      inputs=inputfile.name,
                                                      **options)
             if stderr:
                 raise MetaflowS3Exception("Getting S3 files failed.\n"\
                                           "First prefix requested: %s\n"\
-                                          "Error: %s" % (prefixes[0], stderr))
+                                          "Error: %s" % (prefixes_and_ranges[0], stderr))
             else:
                 for line in stdout.splitlines():
                     yield tuple(map(url_unquote, line.strip(b'\n').split(b' ')))
 
-    def _put_many_files(self, url_files, overwrite):
-        url_files = list(url_files)
+    def _put_many_files(self, url_info, overwrite):
+        url_info = list(url_info)
+        def _update_and_return(d, d1):
```

Comment:
`_update_and_return` is a bit funky construct - making stateful updates etc. Instead, you can replace the line above, `url_info = list(url_info)` with
```
url_info = [dict(chain(('local', os.path.realpath(local)), ('url': url)), info.items()))) for local, url, info in url_info]
```

Reply:
ooh, I like that better. I'll change it.

Reply:
Done